backends/tc/midend.cpp

```diff
@@ -0,0 +1,90 @@
+software distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions
+and limitations under the License.
+*/
+
+#include "midend.h"
+
+namespace TC {
+
+/**
+This class implements a policy suitable for the ConvertEnums pass.
+The policy is: convert all enums that are not part of the v1model.
+Use 32-bit values for all enums.
+*/
+class EnumOn32Bits : public P4::ChooseEnumRepresentation {
+    bool convert(const IR::Type_Enum *type) const override {
+        if (type->srcInfo.isValid()) {
+            return true;
+        }
+        return true;
+    }
+    unsigned enumSize(unsigned) const override { return 32; }
+};
+
+const IR::ToplevelBlock *MidEnd::run(TCOptions &options, const IR::P4Program *program,
+                                     std::ostream *outStream) {
+    if (program == nullptr && options.listMidendPasses == 0) return nullptr;
+    auto evaluator = new P4::EvaluatorPass(&refMap, &typeMap);
+
+    PassManager midEnd = {};
+    midEnd.addPasses(
+        {new P4::ConvertEnums(&refMap, &typeMap, new EnumOn32Bits()),
+         new P4::ClearTypeMap(&typeMap),
+         new P4::RemoveMiss(&refMap, &typeMap),
+         new P4::EliminateInvalidHeaders(&refMap, &typeMap),
+         new P4::EliminateNewtype(&refMap, &typeMap),
+         new P4::SimplifyControlFlow(&refMap, &typeMap),
+         new P4::SimplifyKey(
+             &refMap, &typeMap,
+             new P4::OrPolicy(new P4::IsValid(&refMap, &typeMap), new P4::IsLikeLeftValue())),
+         new P4::RemoveExits(&refMap, &typeMap),
+         new P4::ConstantFolding(&refMap, &typeMap),
+         new P4::SimplifySelectCases(&refMap, &typeMap, false),  // accept non-constant keysets
+         new P4::ExpandEmit(&refMap, &typeMap),
+         new P4::HandleNoMatch(&refMap),
+         new P4::SimplifyParsers(&refMap),
+         new PassRepeated({new P4::ConstantFolding(&refMap, &typeMap),
+                           new P4::StrengthReduction(&refMap, &typeMap)}),
+         new P4::SimplifyComparisons(&refMap, &typeMap),
+         new P4::EliminateTuples(&refMap, &typeMap),
+         new P4::SimplifySelectList(&refMap, &typeMap),
+         new P4::MoveDeclarations(),  // more may have been introduced
+         new P4::RemoveSelectBooleans(&refMap, &typeMap),
+         new P4::SingleArgumentSelect(&refMap, &typeMap),
+         new P4::ConstantFolding(&refMap, &typeMap),
+         new P4::SimplifyControlFlow(&refMap, &typeMap),
+         new P4::TableHit(&refMap, &typeMap),
+         new P4::RemoveLeftSlices(&refMap, &typeMap),
+         new EBPF::Lower(&refMap, &typeMap),
+         new P4::ParsersUnroll(true, &refMap, &typeMap),
+         evaluator,
+         new P4::MidEndLast()});
+    // PSA realted checks
+    midEnd.addPasses(
+        {new P4::ValidateTableProperties({"size", "psa_direct_counter", "psa_direct_meter",
```

Comment:
Some of them are not relevant for the PNA architecture.

Reply:
removed